pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,423 @@
+	if err != nil {
+		return nil, err
+	}
+	if _, err := resp.Body.Read(nil); err != nil {
+		runutil.CloseWithLogOnErr(b.logger, resp.Body, "cos get range obj close")
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	b.opsTotal.WithLabelValues(opObjectHead).Inc()
+	if _, err := b.client.Object.Head(ctx, name, nil); err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "head cos object")
+	}
+
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	tmpErr := toErrorResponse(err)
+	if tmpErr.Code == "NoSuchKey" {
+		return true
+	}
+	if tmpErr.Response != nil && tmpErr.Response.StatusCode == http.StatusNotFound {
+		return true
+	}
+	return false
+}
+
+func toErrorResponse(err error) *cos.ErrorResponse {
+	switch err := err.(type) {
+	case *cos.ErrorResponse:
+		return err
+	default:
+		return &cos.ErrorResponse{}
+	}
+}
+
+func (b *Bucket) Close() error { return nil }
+
+type ObjectInfo struct {
+	cos.Object
+	Err error
+}
+
+func (b *Bucket) ListObjects(ctx context.Context, objectPrefix string, recursive bool) <-chan ObjectInfo {
+	objectsCh := make(chan ObjectInfo, 1)
+	delimiter := "/"
+	if recursive {
+		delimiter = ""
+	}
+
+	go func(objectsCh chan<- ObjectInfo) {
+		defer close(objectsCh)
+		var marker string
+		for {
+			opt := &cos.BucketGetOptions{
+				Prefix: objectPrefix,
+				// TODO(jojohappy) Set this option as prosperity for cos configuration
+				MaxKeys:   1000,
+				Marker:    marker,
+				Delimiter: delimiter,
+			}
+			result, _, err := b.client.Bucket.Get(ctx, opt)
+			if err != nil {
+				objectsCh <- ObjectInfo{
+					Err: err,
+				}
+				return
+			}
+
+			for _, object := range result.Contents {
+				marker = object.Key
+				select {
+				case objectsCh <- ObjectInfo{
+					Object: object,
+					Err:    nil,
+				}:
+				case <-ctx.Done():
+					return
+				}
+			}
+
+			for _, obj := range result.CommonPrefixes {
+				object := cos.Object{}
+				object.Key = obj
+				object.Size = 0
+				select {
+				// Send object prefixes.
+				case objectsCh <- ObjectInfo{
+					Object: object,
+					Err:    nil,
+				}:
+				case <-ctx.Done():
+					return
+				}
+			}
+
+			if result.NextMarker != "" {
+				marker = result.NextMarker
+			}
+
+			if !result.IsTruncated {
```

Comment:
what does it mean `Truncated` in this context? (:

Reply:
Done